File: rust/src/network.rs
use crate::layer::Layer;
use crate::matrix::Matrix;
use crate::optimizer::Optimizer;
use crate::{activation, layer};
use crate::{
    error,
    utils::{random_weights, DataSetItem},
};

pub struct Network {
    error: error::ErrorFunctionKey,
    pub dataset: Vec<DataSetItem>,
    pub layers: Vec<Layer>,
}

pub struct ForwardResult {
    pub results: Vec<layer::ForwardPassResult>,
    pub activated_input: Matrix,
}

pub struct GradientResult {
    pub gradients: Vec<Matrix>,
    pub loss: f64,
}

pub fn create_network(error: error::ErrorFunctionKey) -> Network {
    Network {
        error,
        dataset: Vec::new(),
        layers: Vec::new(),
    }
}

impl Network {
    pub fn set_data(&mut self, dataset: Vec<DataSetItem>) {
        self.dataset = dataset;
    }

    pub fn add(&mut self, activation: activation::ActivationFunctionKey, count: u32) {
        self.layers
            .push(layer::create_layer(count, activation, self.error))
    }

    pub fn compute_gradients(&self) -> GradientResult {
        let mut total_gradient = Vec::new();
        for layer in &self.layers {
            total_gradient.push(layer.weights.scale(0.0))
        }

        let mut total_loss = 0.0;

        for data in &self.dataset {
            let data = self.compute_result(
                &Matrix::from(&data.input),
                &Matrix::from(&data.output),
                &mut total_gradient,
            );
            let results = data.results;
            total_loss += results[results.len() - 1].error / self.dataset.len() as f64
        }

        GradientResult {
            gradients: total_gradient,
            loss: total_loss,
        }
    }

    pub fn compute_result(
        &self,
        real_input: &Matrix,
        expected: &Matrix,
        total: &mut Vec<Matrix>,
    ) -> ForwardResult {
        let data = self.forward_pass(real_input, expected);

        self.backward_pass(&data.activated_input, expected, &data.results, total);

        data
    }

    pub fn forward_pass(&self, input: &Matrix, expected: &Matrix) -> ForwardResult {
        let activated_input = (activation::IDENTITY_ACTIVATION.forward)(&input, false);

        let mut results: Vec<layer::ForwardPassResult> = Vec::new();
        for layer in &self.layers {
            let current_activation = if results.len() > 0 {
                &results[results.len() - 1].activated
            } else {
                &activated_input
            };
            let result = layer.forward_pass(current_activation, &expected);
            results.push(result);
        }

        ForwardResult {
            results,
            activated_input,
        }
    }

    pub fn backward_pass(
        &self,
        input: &Matrix,
        expected: &Matrix,
        results: &Vec<layer::ForwardPassResult>,
        total: &mut Vec<Matrix>,
    ) {
        let result_out = &results[results.len() - 1];
        let layer_out = &self.layers[self.layers.len() - 1];

        let delta_out = layer_out.output_pass(result_out, &expected);
        let mut deltas = vec![delta_out];

        for i in (0..self.layers.len() - 1).rev() {
            let result = &results[i];
            let delta = self.layers[i].backward_pass(&result, &deltas[0], &self.layers[i + 1]);
            deltas.insert(0, delta);
        }

        for l in 0..total.len() {
            let weights = &mut total[l];
            let activations = if l == 0 {
                &input
            } else {
                &results[l - 1].activated
            };
            let delta = &deltas[l];

            for i in 0..weights.rows {
                for j in 0..weights.cols {
                    let diff = delta.get(i, 0) * activations.get(j, 0);
                    let new_val = weights.get(i, j) + (diff / self.dataset.len() as f64);
                    weights.set(i, j, new_val);
                }
            }
        }
    }

    pub fn predict(&self, item: &DataSetItem) -> ForwardResult {
        let input = Matrix::from(&item.input);
        let output = Matrix::from(&item.output);
        self.forward_pass(&input, &output)
    }

    pub fn initialize(&mut self, dataset: Vec<DataSetItem>, weights: Option<Vec<Vec<Vec<f64>>>>) {
        self.dataset = dataset;
        if self.layers.len() == 0 {
            panic!("No layers provided!")
        }
        let mut new_weights: Vec<Vec<Vec<f64>>> = Vec::new();

        let mut current_dim = self.dataset[0].input.len();
        if weights.is_none() {
            for layer in &self.layers {
                let rows = layer.neuron_count as usize;
                let cols = current_dim + 1;

                current_dim = rows;
                new_weights.push(random_weights(rows, cols))
            }
        } else {
            for w in weights.unwrap() {
                new_weights.push(w);
            }
        }

        let count = self.layers.len();
        for l in 0..count {
            let w = &new_weights[l];
            self.layers[l].initialize(Matrix::from(w), l == count - 1);
        }
    }
}

File: javascript/src/Network.ts
import { ActivationFunctionKey, Identity } from "./Activation";
import { Dataset, DatasetItem } from "./Dataset";
import { ErrorFunctionKey } from "./Error";
import { ForwardPassResult, Layer } from "./Layer";
import { Matrix } from "./Matrix";
import { isCorrectCategory, randomWeights } from "./utils";

export type SetResult = {
  loss: number;
  percentage: number;
};

export class Network {
  error: ErrorFunctionKey;
  currentBatch: DatasetItem[];
  layers: Layer[] = [];

  trainSet: Dataset;
  testSet: Dataset | null;

  constructor(
    trainSet: Dataset,
    testSet: Dataset | null,
    error: ErrorFunctionKey
  ) {
    this.trainSet = trainSet;
    this.testSet = testSet;
    this.currentBatch = this.trainSet.generateBatch();
    this.error = error;
  }

  setBatch = (batch: DatasetItem[]) => {
    this.currentBatch = batch;
  };

  /**
   * Calculates forward and backward passes for each item in the currentBatch.
   *   - then calculates the total-loss and the gradient for weights.
   */
  computeGradient = () => {
    // Initialize empty gradient-matrices.
    const totalGradient = this.layers.map((layer) => layer.weights.scale(0));
    let totalLoss = 0;

    const batchLen = this.currentBatch.length;

    // Loop items in batch.
    const results = this.currentBatch.map(({ input, output: expected }) => {
      const data = this.computeResult(
        Matrix.fromList(input),
        Matrix.fromList(expected)
      );
      const { activatedInput, deltas, loss, results } = data;

      // Update total gradient.
      for (let layer = 0; layer < totalGradient.length; layer += 1) {
        const weights = totalGradient[layer];
        const activations =
          layer === 0 ? activatedInput : results[layer - 1].activated;
        const delta = deltas[layer];

        for (let i = 0; i < weights.M; i += 1) {
          for (let j = 0; j < weights.N; j += 1) {
            const diff = (delta.get(i, 0) * activations.get(j, 0)) / batchLen;
            const newValue = weights.get(i, j) + diff;
            weights.set(i, j, newValue);
          }
        }
      }

      // Update total-loss.
      totalLoss += loss / batchLen;

      return {
        input,
        output: expected,
        ...data,
      };
    });

    return {
      results,
      gradient: totalGradient,
      loss: totalLoss,
    };
  };

  /**
   * Calculates the backward and forward pass for a given input-output pair.
   */
  computeResult = (realInput: Matrix, expected: Matrix) => {
    // Calculates neuron-sums and activations for each layer.
    const data = this.forwardPass(realInput, expected);

    // Backropagate
    const deltas = this.backwardPass(expected, data.results);

    return {
      ...data,
      deltas,
    };
  };

  forwardPass = (input: Matrix, expected: Matrix) => {
    // Adds constant 1 to input. Grows dimension by 1 to account for bias term in weight-matrice.
    const activatedInput = Identity.forward(input, false);

    // Keep track of results for each layer.
    const results: ForwardPassResult[] = [];

    let currentActivation = activatedInput;
    for (let i = 0; i < this.layers.length; i += 1) {
      const layer = this.layers[i];
      const result = layer.forwardPass(currentActivation, expected);
      currentActivation = result.activated;
      results.push(result);
    }

    const { activated, error } = results[results.length - 1];
    return {
      loss: error,
      results,
      estimate: activated,
      activatedInput,
    };
  };

  backwardPass = (expected: Matrix, results: ForwardPassResult[]) => {
    const resultOut = results[results.length - 1];

    const layerOut = this.layers[this.layers.length - 1];

    const deltaOut = layerOut.outputPass(resultOut, expected);

    let deltas = [deltaOut];
    // Loop layers in reverse order starting from the second to last layer.
    //  - note that the next-delta is the first element in the deltas-array.
    for (let i = this.layers.length - 2; i >= 0; i -= 1) {
      const result = results[i];

      const delta = this.layers[i].backwardPass(
        result,
        deltas[0],
        this.layers[i + 1]
      );
      deltas.unshift(delta);
    }

    return deltas;
  };

  /**
   * Returns the forward-pass-result for an input-output pair.
   *  - contains estimate and the loss for the estimate.
   */
  predict({ input, output }: DatasetItem) {
    return this.forwardPass(Matrix.fromList(input), Matrix.fromList(output));
  }

  /**
   * Calculates the total-loss and prediction percentage for the whole dataset.
   */
  validateDataset(data: DatasetItem[]): SetResult {
    let totalLoss = 0;
    let correctCount = 0;

    const error = this.error;

    data.forEach((point) => {
      const { loss, estimate } = this.predict(point);
      totalLoss += loss;

      if (error === "log-loss") {
        const result = Math.round(estimate.get(0, 0));
        if (result === point.output[0]) {
          correctCount += 1;
        }
      } else if (error === "cross-entropy") {
        if (isCorrectCategory(estimate, point.output)) {
          correctCount += 1;
        }
      }
    });
    return {
      loss: totalLoss / data.length,
      percentage: correctCount / data.length,
    };
  }

  /**
   * Adds a new layer to the network.
   */
  add(activation: ActivationFunctionKey, neuronCount: number) {
    this.layers.push(new Layer(activation, neuronCount, this.error));
  }

  /**
   * Initializes weights in the network. The weights can also given directly.
   */
  initialize(weights?: number[][][]) {
    if (!this.layers.length) throw new Error("You need to specify layers");
    if (weights && weights.length !== this.layers.length) {
      throw new Error(
        `Invalid amount of weights supplied. Should be ${this.layers.length}.`
      );
    }

    // First dimension is based on the first input of the dataset.
    let currentDimension = this.currentBatch[0].input.length;
    this.layers.map((layer, l) => {
      const m = layer.neuronCount;
      const n = currentDimension + 1;

      currentDimension = layer.neuronCount;

      let w = weights?.[l];
      if (w) {
        if (w.length !== m || w[0].length !== n) {
          throw new Error(
            `Invalid weight dimension at layer ${l}. Should be ${m}x${n}.`
          );
        }
      } else {
        w = randomWeights(m, n);
      }
      layer.initialize(Matrix.fromArrays(w), l === this.layers.length - 1);
    });
  }
}
